dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/MetadataDumper.java

```diff
@@ -255,50 +285,60 @@ public <T> T runChildTask(Task<T> task) throws MetadataDumperUsageException {
 
       } finally {
         // We must do this in finally after the ZipFileSystem has been closed.
-        if (outputFile.isFile()) LOG.debug("Dumper wrote " + outputFile.length() + " bytes.");
+        File outputFile = new File(outputFileName);
+        if (outputFile.isFile()) {
+          LOG.debug("Dumper wrote " + outputFile.length() + " bytes.");
+        }
         LOG.debug("Dumper took " + stopwatch + ".");
       }
 
       printTaskResults(state);
-      printDumperSummary(connector, outputFile);
-      checkRequiredTaskSuccess(state, outputFile);
+      printDumperSummary(connector, outputFileName);
+      checkRequiredTaskSuccess(state, outputFileName);
       logStatusSummary(state);
       System.out.println(STARS);
     }
   }
 
-  private File getOutputFile(Connector connector, ConnectorArguments arguments) {
+  private String getOutputFileName(Connector connector, ConnectorArguments arguments) {
     // The default output file is based on the connector.
     // We had a customer request to base it on the database, but that isn't well-defined,
     // as there may be 0 or N databases in a single file.
     String defaultFileName = connector.getDefaultFileName(arguments.isAssessment());
     return arguments
         .getOutputFile()
         .map(file -> getVerifiedFile(defaultFileName, file))
-        .orElseGet(() -> new File(defaultFileName));
+        .orElseGet(() -> defaultFileName);
   }
 
-  private File getVerifiedFile(String defaultFileName, File file) {
-    String fileName = file.getPath();
-    boolean isZipFile = StringUtils.endsWithIgnoreCase(fileName, ".zip");
+  private String getVerifiedFile(String defaultFileName, String fileName) {
+    Matcher gcsPathMatcher = GCS_PATH_PATTERN.matcher(fileName);
 
-    if (file.isFile() && !isZipFile) {
-      throw new IllegalStateException(
-          String.format(
-              "A file already exists at %1$s. If you want to create a directory, please"
-                  + " provide the path to the directory. If you want to create %1$s.zip,"
-                  + " please add the `.zip` extension manually.",
-              fileName));
+    if (gcsPathMatcher.matches()) {
+      LOG.debug(String.format("Got GCS target with bucket '%s'.", gcsPathMatcher.group("bucket")));
+      boolean isZipFile = StringUtils.endsWithIgnoreCase(fileName, ".zip");
+      return isZipFile ? fileName : fileName + defaultFileName;
```

Comment:
We are dealing with a GCS path here. So, `Path` does not work and we cannot really use the GCS filesystem at this point. I think using file name as a prefix is a reasonable approach since GCS does not really have folders (just an illusion of folders using slashes as a separator - https://cloud.google.com/storage/docs/folders).

Reply:
I am against restructuring the code. It would require moving the creation of the GCS filesystem above the dry-run branch (L238) and having a separate try/catch for it. It would also mean that initial GCS operations are not part of the stopwatch scope anymore.

As for the slash-handling: Allowing it as a prefix could be confusing. Agreed. I added handling to ensure adding a separating /.